common4j/src/main/com/microsoft/identity/common/java/telemetry/events/CertBasedAuthEvent.java

```diff
@@ -0,0 +1,108 @@
+        return this;
+    }
+
+    /**
+     * Puts a Boolean that describes whether or not a PivProvider instance is being removed from the Security static provider list
+     *  at the time of YubiKey removal.
+     * @param isRemoved true when a PivProvider instance is being removed. false otherwise.
+     * @return The Event object.
+     */
+    public CertBasedAuthEvent putPivProviderRemoved(final boolean isRemoved) {
+        put(TelemetryEventStrings.Key.PIVPROVIDER_REMOVED, String.valueOf(isRemoved));
+        return this;
+    }
+
+    /**
+     * Puts a ResultCode received from a RawAuthorizationResult as a result of a completed or failed CBA flow.
+     * @param code ResultCode as a string.
+     * @return The Event object.
+     */
+    public CertBasedAuthEvent putResponseCode(final String code) {
+        put(TelemetryEventStrings.Key.RESULT_CODE, code);
+        return this;
+    }
+
+    /**
+     * Puts details for an Exception received from a RawAuthorizationResult as a result of a failed CBA flow.
+     * Adapted from {@link com.microsoft.identity.common.java.telemetry.events.ErrorEvent#putException(Exception)}.
+     * @param exception BaseException received from RawAuthorizationResult.
+     * @return The Event object.
+     */
+    public CertBasedAuthEvent putResponseException(@NonNull final BaseException exception) {
+        //Code adapted from ErrorEvent.putException
+        final StackTraceElement[] stackTraceElements = exception.getStackTrace();
+        if (stackTraceElements != null && stackTraceElements.length > 0) {
+            final StackTraceElement errorLocation = stackTraceElements[0];
+            final String tag = exception.getClass().getSimpleName() + exception.getMessage() + errorLocation.getClassName() + errorLocation.getMethodName() + exception.getErrorCode();
+
+            // we add Integer.MAX_VALUE to ensure we don't get negatives. This is to ensure consistency and avoid confusion in the tags generated.
+            // For example, if we allow negatives we would generate tags like tag_-123456 and tag_123456, which may seem as similar.
+            final String errorTag = ErrorEvent.ERROR_TAG_PREFIX + (((long) tag.hashCode()) + Integer.MAX_VALUE);
+            put(TelemetryEventStrings.Key.ERROR_TAG, errorTag);
+        }
+
+        put(TelemetryEventStrings.Key.ERROR_CLASS_NAME, exception.getClass().getSimpleName());
+        put(TelemetryEventStrings.Key.ERROR_DESCRIPTION, exception.getMessage()); // pii
+        if (exception.getCause() != null) {
+            put(TelemetryEventStrings.Key.ERROR_CLASS_NAME, exception.getCause().getClass().getSimpleName());
+        }
+
+        put(TelemetryEventStrings.Key.ERROR_CODE, exception.getErrorCode());
+        put(TelemetryEventStrings.Key.SERVER_ERROR_CODE, exception.getCliTelemErrorCode());
+        put(TelemetryEventStrings.Key.SERVER_SUBERROR_CODE, exception.getCliTelemSubErrorCode());
+
+        return this;
```

Comment:
This is all in the ErrorEvent already right? So why duplicate the code here? It seems what you need is to rather emit an ErrorEvent

Reply:
I wanted to keep the response code and exception together in the same cba event, since they're coming from the same object. I guess if it's better to have separate events emitted in order to not duplicate code, that also makes sense to me. 
This would leave just one method then for this event. I'm wondering if it would be better to just emit a BaseEvent instead.

Reply:
Which response code specifically are you talking about?

Reply:
The response code from RawAuthorizationResult: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/9e579b661b4c40326f61d9b5d1049c66e704c717/common4j/src/main/com/microsoft/identity/common/java/providers/RawAuthorizationResult.java#L133

The RawAuthorizationResult comes from here: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/b8ea54b8ea2e37f7cc40939752a6dd582c6ae21c/common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/WebViewAuthorizationFragment.java#L283

Reply:
but where are you using the RawAuthorizationResult over here?

Reply:
The exception comes from the RawAuthorizationResult as well. When the result code corresponds with an error, the RawAuthorizationResult will include an exception (the exception variable will be null otherwise)

Reply:
I meant over here this method just takes Exception and does nothing to RawAuthorizationResult, so I'm not sure what's the relation here. I think my overall question is around why we need this duplicated `putException` method when we have a similar method in ErrorEvent? From your initial response, it seemed like it had something to do with RawAuthorizationResult but I'm not seeing that being used here.

Reply:
The resultCode and the exception both come from RawAuthorizationResult, so I wanted to keep the exception and the corresponding resultCode (exception serving as an explanation for a resultCode) in the same event.
If this doesn't really matter, and it would be better to have the resultCode and the exception in separate events, I'm alright to make that change, too.

Reply:
What question are we looking to answer with the result code? It could definitely be valuable I think but I just want us to enumerate it as a question. 

From code point of view, I think the result code should be in its own event. Something related to the Auth. may be UiEndEvent. 

I think there could be a lot of things that are currently unclear and we may end-up refactoring some of this code in the near term. So I'm not too worried about where we put it right now but we should avoid code duplication.

Reply:
The result code should give an idea as to whether the CBA flow successfully completed or not. This could be helpful in determining the origin of unexpected errors pertaining to security providers.
I'm going to use ErrorEvents when possible and rewrite the CertBasedAuthEvent to be CertBasedAuthResultEvent (This is probably the simplest thing to do right now... I'll refactor when needed). I'll create a separate Event named YubiKeyProviderStatusEvent as well (totally ok with refactoring this later, too).
Hope this plan sounds ok for the time being,